pkg/api/graph/graph.go

```diff
@@ -0,0 +1,397 @@
+package graph
+
+import (
+	"fmt"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/concrete"
+)
+
+type Node struct {
+	concrete.Node
+	UniqueName
+}
+
+type UniqueName string
+
+func (n UniqueName) UniqueName() string {
+	return string(n)
+}
+
+type uniqueNamer interface {
+	UniqueName() string
+}
+
+// UniqueNodeInitializer is a graph that allows nodes with a unique name to be added without duplication.
+// If the node is newly added, true will be returned.
+type UniqueNodeInitializer interface {
+	FindOrCreate(name UniqueName, fn NodeInitializerFunc) (graph.Node, bool)
+}
+
+type NodeInitializerFunc func(Node) graph.Node
+
+func EnsureUnique(g UniqueNodeInitializer, name UniqueName, fn NodeInitializerFunc) graph.Node {
+	node, _ := g.FindOrCreate(name, fn)
+	return node
+}
+
+type MutableDirectedEdge interface {
+	AddEdge(head, tail graph.Node, edgeKind int)
+}
+
+type MutableUniqueGraph interface {
+	graph.Mutable
+	MutableDirectedEdge
+	UniqueNodeInitializer
+}
+
+type Edge struct {
+	concrete.Edge
+	K int
+}
+
+func NewEdge(head, tail graph.Node, kind int) Edge {
+	return Edge{concrete.Edge{head, tail}, kind}
+}
+
+func (e Edge) Kind() int {
+	return e.K
+}
+
+type GraphDescriber interface {
+	Name(node graph.Node) string
+	Kind(node graph.Node) int
+	Object(node graph.Node) interface{}
+	EdgeKind(edge graph.Edge) int
+}
+
+type Interface interface {
```

Comment:
Golang pattern - if your interface is the same as your package name (and you have a concrete object named the same as your package) Interface is the least objectionable form (vs IPackage or similar).

----- Original Message -----
> > +func NewEdge(head, tail graph.Node, kind int) Edge {
> > +	return Edge{concrete.Edge{head, tail}, kind}
> > +}
> > +
> > +func (e Edge) Kind() int {
> > +	return e.K
> > +}
> > +
> > +type GraphDescriber interface {
> > +	Name(node graph.Node) string
> > +	Kind(node graph.Node) int
> > +	Object(node graph.Node) interface{}
> > +	EdgeKind(edge graph.Edge) int
> > +}
> > +
> > +type Interface interface {
> 
> an interface called Interface
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1314/files#r27435072